### PR TITLE
[OpenMPIRBuilder] Added `createTeams`

### PR DESCRIPTION
**Pull request by shraiysh**
This patch adds a generator for the teams construct. The generated IR looks like the following:

```
current_fn() {
  ...
  call @__kmpc_fork_teams(ptr @ident, i32 num_args, ptr @outlined_omp_teams, ...args)
  ...
}
outlined_omp_teams(ptr %global_tid, ptr %bound_tid, ...args) {
  ; teams body
}
```

It does this by first generating the body in the current function. Then we outline the
body in a temporary function. We then create the @outlined_omp_teams function and embed
the temporary outlined function in this function. We then emit the call to runtime
function.

### CODE REVIEW

File: llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h [hunk 1]
```diff
@@ -2005,6 +2005,13 @@ class OpenMPIRBuilder {
   /// \param Loc The insert and source location description.
   void createTargetDeinit(const LocationDescription &Loc);
 
+  /// Generator for `#omp teams`
+  ///
+  /// \param Loc The location where the task construct was encountered.
```
Comment (raghavendhra) [inline llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h:2010]: Nit : task construct -> teams construct

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp [hunk 2]
```diff
@@ -5655,8 +5657,8 @@ GlobalVariable *
 OpenMPIRBuilder::createOffloadMapnames(SmallVectorImpl<llvm::Constant *> &Names,
                                        std::string VarName) {
   llvm::Constant *MapNamesArrayInit = llvm::ConstantArray::get(
-      llvm::ArrayType::get(
-          llvm::PointerType::getUnqual(M.getContext()), Names.size()),
+      llvm::ArrayType::get(llvm::PointerType::getUnqual(M.getContext()),
+                           Names.size()),
```
Comment (jdoerfert) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5661]: It looks like you clang-formated the entire file. Either only format your changes, or pre-commit a NFC that formats the existing file. This way it clutters the changes with unrelated ones.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp [hunk 3]
```diff
@@ -6106,6 +6108,148 @@ void OpenMPIRBuilder::loadOffloadInfoMetadata(Module &M) {
   }
 }
 
+OpenMPIRBuilder::InsertPointTy
+OpenMPIRBuilder::createTeams(const LocationDescription &Loc,
+                             BodyGenCallbackTy BodyGenCB) {
+  if (!updateToLocation(Loc)) {
+    return Loc.IP;
+  }
```
Comment (jdoerfert) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:6116]: Style: No braces

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp [hunk 4]
```diff
@@ -6106,6 +6108,148 @@ void OpenMPIRBuilder::loadOffloadInfoMetadata(Module &M) {
...
+
+    // Create the wrapper function.
+    Builder.SetInsertPoint(StaleCI);
+    SmallVector<Type *> WrapperArgTys{Builder.getPtrTy(), Builder.getPtrTy()};
+    for (auto &Arg : OutlinedFn.args()) {
+      WrapperArgTys.push_back(Arg.getType());
+    }
+    FunctionCallee WrapperFuncVal = M.getOrInsertFunction(
+        "outlined_omp_teams",
+        FunctionType::get(Builder.getVoidTy(), WrapperArgTys, false));
```
Comment (jdoerfert) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:6202]: Where is this function generated? I assume it's internal, if so, we cannot use the name.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp (hunk 2 above)
Comment (shraiysh) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5661]: Oh, sorry. I assumed the file would be formatted. I will update this soon.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp [hunk 5]
```diff
@@ -6106,6 +6108,148 @@ void OpenMPIRBuilder::loadOffloadInfoMetadata(Module &M) {
...
+    assert(IR.isSuccess() && "Inlining outlined omp teams failed");
+    OutlinedFn.eraseFromParent();
+  };
+
+  addOutlineInfo(std::move(OI));
+
+  Builder.SetInsertPoint(ExitBB);
+
+  return Builder.saveIP();
+}
```
[the remark this answers is not in the document]
Comment (shraiysh) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:6251]: This is similar to task handling. Parallel handling uses CodeExtractor to correct the arguments I think. I didn't quite understand that - CodeExtractor and OutlineInfo seem to do similar things.

**Comment by shraiysh** [comment]
Apologies for the ping, unfortunately I have messed up rebasing my branch. I will close this pull request to avoid future notifications for others in this thread. :)

**Pull request closed without merging**